src/resources/platform_handlers/pytorch/model.py

```diff
@@ -0,0 +1,329 @@
+    return ""
+
+
+def _is_py_class_model(model_path):
+    return model_path[-3:] == ".py"
+
+
+def _import_module_from_path(module_name, file_path):
+    spec = importlib.util.spec_from_file_location(module_name, file_path)
+    module = importlib.util.module_from_spec(spec)
+    spec.loader.exec_module(module)
+    return module
+
+
+def _get_model_class_from_module(module):
+    names = dir(module)
+    for name in names:
+        attr = getattr(module, name)
+        try:
+            if issubclass(attr, torch.nn.Module):
+                return attr
+        except TypeError:
+            # attr may not be a class
+            pass
+    raise pb_utils.TritonModelException("Cannot find a subclass of torch.nn.Module")
+
+
+def _parse_io_config(io_config):
+    io = []
+    for conf in io_config:
+        io.append({"name": conf["name"]})
+    return io
+
+
+def _get_device_name(model_instance_kind, model_instance_device_id):
+    if model_instance_kind == "GPU":
+        return "cuda:" + model_instance_device_id
+    if model_instance_kind == "CPU":
+        return "cpu"
+    # unspecified device
+    return ""
+
+
+def _get_device(model_instance_kind, model_instance_device_id):
+    device_name = _get_device_name(model_instance_kind, model_instance_device_id)
+    if device_name != "":
+        return torch.device(device_name)
+    # leave torch objects at its default device
+    return None
+
+
+def _torch_object_to_device(obj, device):
+    if device != None:
+        return obj.to(device)
+    # leave object at its current device
+    return obj
+
+
+def _set_torch_parallelism(config):
+    log_msg = ""
+    parallelism_settings = ["NUM_THREADS", "NUM_INTEROP_THREADS"]
+    for setting in parallelism_settings:
+        torch_setting = setting.lower()
+        if setting in config["parameters"]:
+            val = config["parameters"][setting]["string_value"]
+            getattr(torch, "set_" + torch_setting)(int(val))
+        val = str(getattr(torch, "get_" + torch_setting)())
+        log_msg += setting + " = " + val + "; "
+    return log_msg
+
+
+def _get_torch_compile_params(config):
+    params = {}
+    if "TORCH_COMPILE_OPTIONAL_PARAMETERS" in config["parameters"]:
+        val = config["parameters"]["TORCH_COMPILE_OPTIONAL_PARAMETERS"]["string_value"]
+        params = json.loads(val)
+        if "model" in params:
+            raise pb_utils.TritonModelException(
+                "'model' is not an optional parameter for 'torch.compile'"
+            )
+    return params
+
+
+def _batch_torch_tensors(requests_tensors):
+    batch_tensors = []
+    sections = []
+    for i in range(len(requests_tensors)):
+        request_tensors = requests_tensors[i]
+        for j in range(len(request_tensors)):
+            request_tensor = request_tensors[j]
+            if j < len(batch_tensors):
+                # add to existing tensor
+                tensor = batch_tensors[j]
+                torch.cat((tensor, request_tensor), 0)
```

Comment:
What if the request tensors are on different devices? Would this still work properly?

Reply:
I guess it will not work, but the device info is embedded into each torch tensor, so if it does not work, the error will be internal to the torch framework.

To address the issue, the user will have to specify the model instance kind, which the script will ensure both the model and tensors are on the same device as described by the model instance kind.

Reply:
In other words, this can only happen if the `KIND_MODEL` is used.

Reply:
I mean let's say a tensor from request A is stored on GPU and tensor from request B is on CPU, is torch.cat able to concatenate them into a single tensor? I believe torch was sensitive for all the tensors to have the same type so just wanted to make sure.

Reply:
> I mean let's say a tensor from request A is stored on GPU and tensor from request B is on CPU, is torch.cat able to concatenate them into a single tensor?

I do not think this will work.

If the user uses `KIND_GPU` or `KIND_CPU`, the input tensors will be moved to either CPU or GPU before `gather` is called (before `torch.cat()` is called), so all tensors will be on the same device. However, if the user uses `KIND_MODEL`, the tensors will not be moved, which could cause the issue. To be safe, the user should avoid using `KIND_MODEL`.

Reply:
I gave it a try: Place model parameters on GPU, and place tensors on CPU. It does not work. The model parameters and tensors must be on the same device.

How about we do it this way, for `KIND_MODEL`:

If the model has at least one parameter, the first parameter of the model `next(model.parameters()).device` will be used to determine the model device. The entire model (i.e. any subsequence parameters not on the same device as the first parameter) will be moved to the model device determined. Then, any input tensors will be moved to the model device.

If no parameter is specified for the model, input tensors will be unified at `cpu`.

Reply:
[Update kind_model behavior](https://github.com/triton-inference-server/python_backend/pull/282/commits/f0f03b9c10e112fb332c2c3507321ba041199497)

Reply:
[Simplify kind_model logic](https://github.com/triton-inference-server/python_backend/pull/282/commits/9cf3041720d9f700465ec290e6929ecd8c04061d)